sw/otbn/crypto/p384_curve_point_valid.s

```diff
@@ -0,0 +1,204 @@
+ * is invalid.
+ *
+ * @param[in]  dmem[12]: dptr_r, pointer to dmem location where right
+ *                               side result r will be stored
+ * @param[in]  dmem[16]: dptr_s, pointer to dmem location where left side
+ *                               result s will be stored
+ * @param[in]  dmem[20]: dptr_x, pointer to dmem location containing affine
+ *                               x-coordinate of input point
+ * @param[in]  dmem[24]: dptr_y, pointer to dmem location containing affine
+ *                               y-coordinate of input point
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * clobbered registers: x2, x3, x20 to x23, w0 to w17
+ * clobbered flag groups: FG0
+ */
+p384_curve_point_valid:
+  /* Init all-zero register. */
+  bn.xor    w31, w31, w31
+
+  /* load domain parameter p (modulus)
+     [w13, w12] = p = dmem[p384_p] */
+  li        x2, 12
+  la        x3, p384_p
+  bn.lid    x2++, 0(x3)
+  bn.lid    x2++, 32(x3)
+
+  /* Load public key x-coordinate.
+     [w11, w10] <= dmem[x] = x */
+  la        x20, dptr_x
+  lw        x20, 0(x20)
+  li        x2, 10
+  bn.lid    x2++, 0(x20)
+  bn.lid    x2, 32(x20)
+
+  /* Compare x to p.
+       FG0.C <= (x < p) */
+  bn.sub    w0, w10, w12
+  bn.subb   w0, w11, w13
+
+  /* Trigger a fault if FG0.C is false. */
+  csrrs     x2, 0x7c0, x0
+  andi      x2, x2, 1
+  bne       x2, x0, _x_valid
+  unimp
+
+  _x_valid:
+
+  /* Load public key y-coordinate.
+       w2 <= dmem[y] = y */
+  la        x21, dptr_y
+  lw        x21, 0(x21)
+  li        x2, 8
+  bn.lid    x2++, 0(x21)
+  bn.lid    x2, 32(x21)
+
+  /* Compare y to p.
+       FG0.C <= (y < p) */
+  bn.sub    w0, w8, w12
+  bn.subb   w0, w9, w13
+
+  /* Trigger a fault if FG0.C is false. */
+  csrrs     x2, 0x7c0, x0
+  andi      x2, x2, 1
+  bne       x2, x0, _y_valid
+  unimp
+
+  _y_valid:
+
+  /* Compute both sides of the Weierstrauss equation.
+       dmem[r] <= (x^3 + ax + b) mod p
+       dmem[s] <= (y^2) mod p */
+  jal       x1, p384_isoncurve
+
+  /* Load both sides of the equation.
+       [w7, w6] <= dmem[r]
+       [w5, w4] <= dmem[s] */
+  la        x22, dptr_r
+  lw        x22, 0(x22)
+  li        x2, 6
+  bn.lid    x2++, 0(x22)
+  bn.lid    x2, 32(x22)
+  la        x23, dptr_s
+  lw        x23, 0(x23)
+  li        x2, 4
+  bn.lid    x2++, 0(x23)
+  bn.lid    x2, 32(x23)
+
+  /* Compare the two sides of the equation.
+       FG0.Z <= (y^2) mod p == (x^2 + ax + b) mod p */
+  bn.sub    w0, w4, w6
+  bn.subb   w1, w5, w7
+
+  bn.cmp    w0, w31
+
+  /* Trigger a fault if FG0.Z is false. */
+  csrrs     x2, 0x7c0, x0
+  srli      x2, x2, 3
+  andi      x2, x2, 1
+  bne       x2, x0, _pt_1st_reg_valid
+  unimp
+
+  _pt_1st_reg_valid:
+
+  bn.cmp    w1, w31
+
+  /* Trigger a fault if FG0.Z is false. */
+  csrrs     x2, 0x7c0, x0
+  srli      x2, x2, 3
+  andi      x2, x2, 1
+  bne       x2, x0, _pt_valid
+  unimp
+
+  _pt_valid:
```

Comment:
Maybe move the jump above, or add more `unimp` repetitions, as an FI protection?

Reply:
Good point, but wouldn't it then also make sense to add more `unimp` instructions after the above jumps?

Reply:
You can, but I think these are less critical from a security standpoint as long as (x, y) still checks out as on-curve.